lib/rtti.h

```diff
@@ -0,0 +1,181 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef LIB_RTTI_H_
+#define LIB_RTTI_H_
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <type_traits>
+
+namespace RTTI {
+
+using TypeId = std::uint64_t;
+static constexpr TypeId InvalidTypeId = UINT64_C(0);
+
+namespace detail {
+// Apparently string_view does not optimize properly in constexpr context
+// for gcc < 13
+struct TypeNameHolder {
+    template <size_t N>
+    constexpr TypeNameHolder(const char (&str)[N]) : str(&str[0]), length(N - 1) {}  // NOLINT
+
+    const char *str;
+    size_t length;
+};
+
+template <typename T>
+constexpr TypeNameHolder TypeNameProxy() {
+#ifdef __clang__
+    return __PRETTY_FUNCTION__;
+#elif defined(__GNUC__)
+    return __PRETTY_FUNCTION__;
+#else
+#error "Unsupported compiler"
+#endif
+}
+
+static constexpr uint64_t FNV1a(const char *str, size_t n,
```

Comment:
Note: There is already `fnv1a` in `lib/hash.h`, but looking at it, its interface is quite a bit different so adding this version alongside it could be more confusing then helpful (here we do need `uint64_t` output, while the one in `hash.h` uses `size_t` and works also in cases where `size_t` is 32 bits).

Reply:
Yes, this is an explicit 64-bit FNV1a implementation, not 32-bit one.

Reply:
This is something we should keep track off. I think the use of that hash function is limited to the cstring implementation.

Reply:
@fruffy Funny enough, the situation is a bit more complicated. `lib/hash.cpp` implements fnv1a and (old) murmur2. murmur2 is used in `cstring` (for unification, as string hash) and fnv1 is used to hash cstring' contents insider parser unroll. Both cases do not make much sense to me:

1. murmur2 is old, slow and known to have flaws
2. fnv1a is quite slow, has mediocre quality. And actually its usage in parser unroll is questionable: since cstrings are guaranteed unique, then the address of string is enough as 64-bit hash

In my opinion:
 - murmur2 should be replaced with something a bit more modern and performant. I'd stick with xxhash3.
 - fnv1a usage from parser unroll should be removed

The only reason why fnv1a was chosen here is that it's small enough to be evaluated in constexpr context without slowing down compilation time. And for this its quality is good enough here: note that we are hashing few dozen of strings that are known during compile time, so our collision requirements are much lower than generic purpose string hash.

PS: Certainly, hash refactoring is a separate task :)

Reply:
I will create an issue for this. 

`ParsersUnroll` is a troubled compiler pass. We shouldn't shy away from making improvements to it and replacing the hash function, if needed.